src/fsharp/service/IncrementalBuild.fs

```diff
@@ -1041,27 +1117,323 @@ type TypeCheckAccumulator =
+            | None, _ -> true
+            | Some(PartialState _), false -> true
+            | _ -> false
+
+        if mustCheck then
+            lazyTcInfoState := None
+
+        match !lazyTcInfoState with
+        | Some tcInfoState -> tcInfoState |> Eventually.Done
+        | _ -> 
+            eventually {
+                let! tcInfoState = this.TypeCheck(partialCheck)
+                lazyTcInfoState := Some tcInfoState
+                return tcInfoState
+            }
+
+    member this.Next(syntaxTree) =
+        eventually {
+            let! prevState = this.GetState(true)
+            let lazyPrevTcInfoOptional =
+                eventually {
+                    let! prevState = this.GetState(false)
+                    match prevState with
+                    | FullState(_, prevTcInfoOptional) -> return Some prevTcInfoOptional
+                    | _ -> return None
+                }
+            return
+                SemanticModel(
+                    tcConfig,
+                    tcGlobals,
+                    tcImports,
+                    keepAssemblyContents, 
+                    keepAllBackgroundResolutions, 
+                    maxTimeShareMilliseconds, 
+                    keepAllBackgroundSymbolUses, 
+                    enableBackgroundItemKeyStoreAndSemanticClassification,
+                    enablePartialTypeChecking,
+                    beforeFileChecked, 
+                    fileChecked, 
+                    prevState.Partial, 
+                    lazyPrevTcInfoOptional, 
+                    Some syntaxTree,
+                    ref None)
+        }
+
+    member this.Finish(finalTcErrorsRev, finalTopAttribs) =
+        eventually {
+            let! state = this.GetState(true)
+
+            let finishTcInfo = { state.Partial with tcErrorsRev = finalTcErrorsRev; topAttribs = finalTopAttribs }
+            let finishState =
+                match state with
+                | PartialState(_) -> PartialState(finishTcInfo)
+                | FullState(_, tcInfoOptional) -> FullState(finishTcInfo, tcInfoOptional)
+
+            return
+                SemanticModel(
+                    tcConfig,
+                    tcGlobals,
+                    tcImports,
+                    keepAssemblyContents, 
+                    keepAllBackgroundResolutions, 
+                    maxTimeShareMilliseconds, 
+                    keepAllBackgroundSymbolUses, 
+                    enableBackgroundItemKeyStoreAndSemanticClassification,
+                    enablePartialTypeChecking,
+                    beforeFileChecked, 
+                    fileChecked, 
+                    prevTcInfo, 
+                    prevTcInfoOptional, 
+                    syntaxTreeOpt,
+                    ref (Some finishState))
+        }
+
+    member this.TcInfo =
+        eventually {
+            let! state = this.GetState(true)
+            return state.Partial
+        }
+
+    member this.TcInfoFull =
```

Comment:
it's not clear to me what this name means? the value gets an unnamed tuple of TcInfo, TcInfoOptional.

perhaps the name should be this.TcInfoAndTcInfoOptional ... which is yuk!!! but at least it tells us what we are getting.

oh I don't know.

Reply:
Ah yes, this name is not great, but I'm struggling to figure out what I should call it.

Basically, whenever I try to access this property, it's going to try to type-check the file to get the full and rich information that would be needed for tooling. `TcInfoAndTcInfoOptional` is accurate and might be better though.